spec/02-integration/04-admin_api/24-reconfiguration-completion_spec.lua

```diff
@@ -0,0 +1,156 @@
+local helpers = require "spec.helpers"
+local cjson = require "cjson"
+
+describe("Admin API - Reconfiguration Completion -", function()
+
+  local WORKER_STATE_UPDATE_FREQ = 1
+
+  local admin_client
+  local proxy_client
+
+  local function run_tests()
+
+    local res = admin_client:post("/plugins", {
+      body = {
+        name = "request-termination",
+        config = {
+          status_code = 200,
+          body = "kong terminated the request",
+        }
+      },
+      headers = { ["Content-Type"] = "application/json" },
+    })
+    assert.res_status(201, res)
+
+    res = admin_client:post("/services", {
+      body = {
+        name = "test-service",
+        url = "http://127.0.0.1",
+      },
+      headers = { ["Content-Type"] = "application/json" },
+    })
+    local body = assert.res_status(201, res)
+    local service = cjson.decode(body)
+
+    -- We're running the route setup in `eventually` to cover for the unlikely case that reconfiguration completes
+    -- between adding the route and requesting the path through the proxy path.
+
+    local next_path do
+      local path_suffix = 0
+      function next_path()
+        path_suffix = path_suffix + 1
+        return "/" .. tostring(path_suffix)
+      end
+    end
+
+    local service_path
+    local kong_transaction_id
+
+    assert.eventually(function()
+      service_path = next_path()
+
+      res = admin_client:post("/services/" .. service.id .. "/routes", {
+        body = {
+          paths = { service_path }
+        },
+        headers = { ["Content-Type"] = "application/json" },
+      })
+      assert.res_status(201, res)
+      kong_transaction_id = res.headers['x-kong-transaction-id']
+      assert.is_string(kong_transaction_id)
+
+      res = proxy_client:get(service_path,
+              {
+                headers = {
+                  ["X-If-Kong-Transaction-Id"] = kong_transaction_id
+                }
+              })
+      assert.res_status(503, res)
+      assert.equals("pending", res.headers['x-kong-reconfiguration-status'])
+      local retry_after = tonumber(res.headers['retry-after'])
+      ngx.sleep(retry_after)
```

Comment:
totally optional but should we use the `.with_step()` modifier instead of sleeping?

Reply:
The problem is that the `retry-after` period is only known after the request failed, so `with_step()` would be a bit iffy (its argument would be an up-value that would be set in the first failing call).  I think it is cleaner this way.